### PR TITLE
Add AllowSynchronousIO to TestServer and IIS, fix tests

### PR DESCRIPTION
**Pull request by Tratcher**
#4774 This splits the prior PR (https://github.com/aspnet/AspNetCore/pull/5120) and takes the sync IO improvements without changing the server defaults. We'll followup on the default change once MVC is fixed.

### CODE REVIEW

File: src/Hosting/TestHost/src/AsyncStreamWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+        public override bool CanWrite => _inner.CanWrite;
+
+        public override long Length => _inner.Length;
+
+        public override long Position { get => _inner.Position; set => _inner.Position = value; }
+
+        public override void Flush()
+        {
+            // Not bocking Flush because things like StreamWriter.Dispose() always call it.
```
Comment (jkotalik) [inline src/Hosting/TestHost/src/AsyncStreamWrapper.cs]: nit: blocking

File: src/Hosting/TestHost/src/TestServer.cs [hunk 2]
```diff
@@ -77,6 +77,14 @@ public IWebHost Host
 
         public IFeatureCollection Features { get; }
 
+        /// <summary>
+        /// Gets or sets a value that controls whether synchronous IO is allowed for the <see cref="HttpContext.Request"/> and <see cref="HttpContext.Response"/>
+        /// </summary>
+        /// <remarks>
+        /// Defaults to true.
+        /// </remarks>
+        public bool AllowSynchronousIO { get; set; } = true;
```
Comment (jkotalik) [inline src/Hosting/TestHost/src/TestServer.cs]: Are we eventually going to be able to remove this flag (or default it to false).

**Review by jkotalik — APPROVED** [review]

**Review by davidfowl — APPROVED** [review]

File: src/Security/Authentication/test/CookieTests.cs [hunk 3]
```diff
@@ -1510,7 +1510,7 @@ private static TestServer CreateServerWithServices(Action<IServiceCollection> co
             return server;
         }
 
-        private static void Describe(HttpResponse res, AuthenticateResult result)
+        private static Task Describe(HttpResponse res, AuthenticateResult result)
```
Comment (JunTaoLuo) [inline src/Security/Authentication/test/CookieTests.cs]: nit: DescribeAsync

File: src/Hosting/TestHost/src/AsyncStreamWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.TestHost
+{
+    internal class AsyncStreamWrapper : Stream
+    {
+        private Stream _inner;
+        private Func<bool> _allowSynchronousIO;
+
+        internal AsyncStreamWrapper(Stream inner, Func<bool> allowSynchronousIO)
```
Comment (JunTaoLuo) [inline src/Hosting/TestHost/src/AsyncStreamWrapper.cs]: Is the reason for the Func so the value is checked on every operation between which the value could be changed?

**Review by JunTaoLuo — APPROVED** [review]

File: src/Hosting/TestHost/src/TestServer.cs (hunk 2 above)
Comment (Tratcher) [reply src/Hosting/TestHost/src/TestServer.cs]: Defaulting to false is coming in the next PR (https://github.com/aspnet/AspNetCore/pull/5120) once we deal with MVC.

File: src/Hosting/TestHost/src/AsyncStreamWrapper.cs (hunk 4 above)
Comment (Tratcher) [reply src/Hosting/TestHost/src/AsyncStreamWrapper.cs]: Right, the per request feature could be changed after the stream was created.

**Pull request merged**